Gems/Atom/Asset/Shader/Code/Source/Editor/CommonFiles/Preprocessor.cpp

```diff
@@ -391,24 +431,67 @@ namespace AZ
 
+            bool useDefaultProjectShaderLibFromEngine = true;
+
+            AZ::IO::Path enginePath{ AZStd::string_view{ AZ::Utils::GetEnginePath() } };
+
             // but while we transfer to the set, we're going to keep only folders where +/ShaderLib exists
-            for (AZ::IO::Path shaderScanFolder : scanFoldersVector)
+            for (AZ::IO::Path inShaderScanFolder : scanFoldersVector)
             {
-                shaderScanFolder /= "ShaderLib";
+                auto shaderScanFolder = inShaderScanFolder / "ShaderLib";
                 if (auto it = AZStd::find_if(includePaths.begin(), includePaths.end(), PathCompare(shaderScanFolder));
                     it == includePaths.end())
                 {
                     // the folders constructed this fashion constitute the base of automatic include search paths
                     if (AZ::IO::SystemFile::Exists(shaderScanFolder.c_str()))
                     {
-                        includePaths.emplace_back(AZStd::move(shaderScanFolder.LexicallyNormal().Native()));
+                        if (inShaderScanFolder == projectPath)
+                        {
+                            // use the <project>/ShaderLib - folder only if contents were modified. 
+                            // Otherwise use the DefaultProjectShaderlib folder from the engine
+                            if (FilesModifiedFromTemplate(shaderScanFolder)) {
+                                AZ_TracePrintf(
+                                    "Preprocessing",
+                                    "The files in %s/ShaderLib were modified since initial creation, and have priority over "
+                                    "%s/Gems/Atom/Asset/DefaultProjectShaderlib/. Be aware that "
+                                    "potential upstream changes need to be migrated to the project manually.",
+                                    projectPath.c_str(),
+                                    enginePath.c_str());
+                                useDefaultProjectShaderLibFromEngine = false;
+                                includePaths.emplace_back(AZStd::move(shaderScanFolder.LexicallyNormal().Native()));
+                            }
+                            else {
+                                AZ_TracePrintf(
+                                    "Preprocessing",
+                                    "The files in %s/ShaderLib are unmodified since initial creation: using the versions from "
+                                    "%s/Gems/Atom/Asset/DefaultProjectShaderlib instead. Consider removing the folder from the "
+                                    "project.",
+                                    projectPath.c_str(),
+                                    enginePath.c_str());
+                                continue;
+                            }
+                        }
+                        else
```

Comment:
I think code would be cleaner by removing this else condition (and removing includePaths line 461 as a result)